content/en/blog/_posts/2025-02-14-cloud-controller-manager-chicken-egg-problem.md

```diff
@@ -0,0 +1,124 @@
+This migration was complex and some logic had to be extracted from the core components, building four new subsystems.
+
+1. **Cloud controller manager** ([KEP-2392](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/2392-cloud-controller-manager/README.md))  
+2. **API server network proxy** ([KEP-1281](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1281-network-proxy))  
+3. **kubelet credential provider plugins** ([KEP-2133](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2133-kubelet-credential-providers))  
+4. **Storage migration to use [CSI](https://github.com/container-storage-interface/spec?tab=readme-ov-file#container-storage-interface-csi-specification-)** ([KEP-625](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/625-csi-migration/README.md))
+
+The [cloud controller manager is part of the control plane](https://kubernetes.io/docs/concepts/architecture/cloud-controller/). It is a critical component that replaces some functionality that existed previously in the kube-controller-manager and the kubelet.
+
+![Components of Kubernetes](https://kubernetes.io/images/docs/components-of-kubernetes.svg)
+
+One of the most critical functionalities of the cloud controller manager is the node controller, which is responsible for the initialization of the nodes.
+
+As we can see in the following diagram, when the **kubelet** starts, it registers the `Node` object with the apiserver, Tainting the node so it can be processed first by the cloud-controller-manager. The initial `Node` is missing the cloud-provider specific information, like the Node Addresses and the Labels with the cloud provider specific information like the Node, Region and Instance type information.
+
+![Diagram of Node initialization](https://mermaid.ink/img/pako:eNqVU01PAjEQ_StNTyrLChgObgyJwYtRiPEjXvZStsPSuNvitMUg4b87u-VjIRLjHjad9nXemzfTFc-MBJ5wC58edAZ3SuQoylQz-oR3RvtyAhhihMwxzCdn3etuxHq9K_r1--fh8MFPoADXHgxa1bIt5soCLgATNkQQDtiYqAJ2bCg0dLa9lbBXobRLbiZ4OWCagHFWGC_naBZKAsYfhEMNDmyszJ5wz0K87V2yioltaGVAg5a_cTdl1reUrRDsGYRcBjW1MpARGylrlc4D7lZKBGvBXkQsjuO_cr8QP3ubSxJkj8zsdToR6_Y7x2Ye1NYaVna0M6MdmqKgvVJokVep34XLZqQevoKyg7L3ik4nuNfKKVGo79Ci0IQazp42_rNHQdba6Kj2Rumn0m_70jQj-NCYh380p25MqnnES8BSKEnDu6oup9zNoISUJ7SUMBW-cClP9Zqg1SC_LHXGE4ceIo7G5zOeTEVhKfK1nM3k73ZBKmdwFJ5H_UrWPzszEMk?type=png\)]\(https://mermaid.live/edit#pako:eNqVU01PAjEQ_StNTyrLChgObgyJwYtRiPEjXvZStsPSuNvitMUg4b87u-VjIRLjHjad9nXemzfTFc-MBJ5wC58edAZ3SuQoylQz-oR3RvtyAhhihMwxzCdn3etuxHq9K_r1--fh8MFPoADXHgxa1bIt5soCLgATNkQQDtiYqAJ2bCg0dLa9lbBXobRLbiZ4OWCagHFWGC_naBZKAsYfhEMNDmyszJ5wz0K87V2yioltaGVAg5a_cTdl1reUrRDsGYRcBjW1MpARGylrlc4D7lZKBGvBXkQsjuO_cr8QP3ubSxJkj8zsdToR6_Y7x2Ye1NYaVna0M6MdmqKgvVJokVep34XLZqQevoKyg7L3ik4nuNfKKVGo79Ci0IQazp42_rNHQdba6Kj2Rumn0m_70jQj-NCYh380p25MqnnES8BSKEnDu6oup9zNoISUJ7SUMBW-cClP9Zqg1SC_LHXGE4ceIo7G5zOeTEVhKfK1nM3k73ZBKmdwFJ5H_UrWPzszEMk)
```

Comment:
@elmiko For posterity, could we put this image in as part of the PR?

If I remember correctly, mermaid.js is also available on both k8s.io and k8s.dev, too.

Reply:
absolutely